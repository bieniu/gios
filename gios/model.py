"""Type definitions for GIOS."""
from dataclasses import dataclass
from typing import Optional, Union


@dataclass
class Sensor:
    """Data class for sensor."""

    name: str
    id: Optional[int]
    index: Optional[str] = None
    value: Optional[Union[float, str]] = None


@dataclass
class GiosSensors:
    """Data class for polutants."""

    aqi: Optional[Sensor]
    c6h6: Optional[Sensor]
    co: Optional[Sensor]
    no2: Optional[Sensor]
    o3: Optional[Sensor]
    pm10: Optional[Sensor]
    pm25: Optional[Sensor]
    so2: Optional[Sensor]
