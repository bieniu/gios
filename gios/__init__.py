"""
Python wrapper for getting air quality data from GIOS.
"""
import logging

_LOGGER = logging.getLogger(__name__)

ATTR_AQI = "AQI"
ATTR_ID = "id"
ATTR_INDEX = "index"
ATTR_INDEX_LEVEL = "{}IndexLevel"
ATTR_NAME = "name"
ATTR_VALUE = "value"

HTTP_OK = 200
URL_INDEXES = "http://api.gios.gov.pl/pjp-api/rest/aqindex/getIndex/{}"
URL_SENSOR = "http://api.gios.gov.pl/pjp-api/rest/data/getData/{}"
URL_STATION = "http://api.gios.gov.pl/pjp-api/rest/station/sensors/{}"
URL_STATIONS = "http://api.gios.gov.pl/pjp-api/rest/station/findAll"


class Gios:  # pylint:disable=too-many-instance-attributes
    """Main class to perform GIOS API requests"""

    def __init__(self, station_id, session):
        """Initialize."""
        self.data = {}
        self.station_id = station_id
        self.latitude = None
        self.longitude = None
        self.station_name = None
        self._station_data = {}
        self._available = False

        self.session = session

    async def update(self):  # pylint:disable=too-many-branches
        """Update GIOS data."""
        data = {}
        invalid_sensors = []

        if not self.station_name:
            stations = await self._get_stations()
            if not stations:
                self._available = False
                raise ApiError("Invalid measuring stations list from GIOS API")

            for station in stations:
                if station[ATTR_ID] == self.station_id:
                    self.latitude = station["gegrLat"]
                    self.longitude = station["gegrLon"]
                    self.station_name = station["stationName"]
            if not self.station_name:
                self._available = False
                raise NoStationError(
                    f"{self.station_id} is not a valid measuring station ID"
                )

            self._station_data = await self._get_station()

        if not self._station_data:
            self._available = False
            raise InvalidSensorsData("Invalid measuring station data from GIOS API")

        for sensor in self._station_data:
            data[sensor["param"]["paramCode"]] = {
                ATTR_ID: sensor[ATTR_ID],
                ATTR_NAME: sensor["param"]["paramName"],
            }

        sensors = await self._get_all_sensors(data)

        # The GIOS server sends a null values for sensors several minutes before
        # adding new data from measuring station. If the newest value is null
        # we take the earlier value.
        for sensor in data:
            try:
                if sensors[sensor]["values"][0][ATTR_VALUE]:
                    data[sensor][ATTR_VALUE] = sensors[sensor]["values"][0][ATTR_VALUE]
                elif sensors[sensor].get("values")[1][ATTR_VALUE]:
                    data[sensor][ATTR_VALUE] = sensors[sensor]["values"][1][ATTR_VALUE]
                else:
                    invalid_sensors.append(sensor)
            except (IndexError, KeyError, TypeError):
                invalid_sensors.append(sensor)

        if invalid_sensors:
            for sensor in invalid_sensors:
                data.pop(sensor)

        if not data:
            raise InvalidSensorsData("Invalid sensor data from GIOS API")

        indexes = await self._get_indexes()

        try:
            for sensor in data:
                index_level = ATTR_INDEX_LEVEL.format(sensor.lower().replace(".", ""))
                data[sensor][ATTR_INDEX] = indexes[index_level][
                    "indexLevelName"
                ].lower()

            data[ATTR_AQI] = {ATTR_NAME: ATTR_AQI}
            data[ATTR_AQI][ATTR_VALUE] = indexes["stIndexLevel"][
                "indexLevelName"
            ].lower()
        except (IndexError, KeyError, TypeError) as err:
            self._available = False
            raise InvalidSensorsData("Invalid index data from GIOS API") from err
        self._available = True
        self.data = data

    async def _get_stations(self):
        """Retreive list of measuring stations."""
        stations = await self._async_get(URL_STATIONS)
        return stations

    async def _get_station(self):
        """Retreive measuring station data."""
        url = URL_STATION.format(self.station_id)
        station = await self._async_get(url)
        return station

    async def _get_all_sensors(self, sensors):
        """Retreive all sensors data."""
        data = {}
        for sensor in sensors:
            sensor_data = await self._get_sensor(sensors[sensor][ATTR_ID])
            data[sensor] = sensor_data
        return data

    async def _get_sensor(self, sensor):
        """Retreive sensor data."""
        url = URL_SENSOR.format(sensor)
        sensor = await self._async_get(url)
        return sensor

    async def _get_indexes(self):
        """Retreive indexes data."""
        url = URL_INDEXES.format(self.station_id)
        indexes = await self._async_get(url)
        return indexes

    async def _async_get(self, url):
        """Retreive data from GIOS API."""
        data = None
        async with self.session.get(url) as resp:
            if resp.status != HTTP_OK:
                _LOGGER.warning("Invalid response from GIOS API: %s", resp.status)
                self._available = False
                raise ApiError(resp.status)
            data = await resp.json()
        _LOGGER.debug("Data retrieved from %s, status: %s", url, resp.status)
        return data

    @property
    def available(self):
        """Return True is data is available."""
        return bool(self.data)


class ApiError(Exception):
    """Raised when GIOS API request ended in error."""

    def __init__(self, status):
        """Initialize."""
        super().__init__(status)
        self.status = status


class InvalidSensorsData(Exception):
    """Raised when sensors data is invalid."""

    def __init__(self, status):
        """Initialize."""
        super().__init__(status)
        self.status = status


class NoStationError(Exception):
    """Raised when no measuring station error."""

    def __init__(self, status):
        """Initialize."""
        super().__init__(status)
        self.status = status
