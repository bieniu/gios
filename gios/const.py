"""Constants for GIOS library."""
from typing import Final

ATTR_AQI: Final[str] = "AQI"
ATTR_ID: Final[str] = "id"
ATTR_INDEX: Final[str] = "index"
ATTR_INDEX_LEVEL: Final[str] = "{}IndexLevel"
ATTR_NAME: Final[str] = "name"
ATTR_VALUE: Final[str] = "value"

URL_INDEXES: Final[str] = "http://api.gios.gov.pl/pjp-api/rest/aqindex/getIndex/{}"
URL_SENSOR: Final[str] = "http://api.gios.gov.pl/pjp-api/rest/data/getData/{}"
URL_STATION: Final[str] = "http://api.gios.gov.pl/pjp-api/rest/station/sensors/{}"
URL_STATIONS: Final[str] = "http://api.gios.gov.pl/pjp-api/rest/station/findAll"
