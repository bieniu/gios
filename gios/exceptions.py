"""GIOS exceptions."""


class GiosError(Exception):
    """Base class for GIOS errors."""

    def __init__(self, status: str) -> None:
        """Initialize."""
        super().__init__(status)
        self.status = status


class ApiError(GiosError):
    """Raised when GIOS API request ended in error."""


class InvalidSensorsDataError(GiosError):
    """Raised when sensors data is invalid."""


class NoStationError(GiosError):
    """Raised when no measuring station error."""
