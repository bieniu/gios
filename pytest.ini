[pytest]
addopts=--cov --cov-report term-missing --disable-pytest-warnings --error-for-skips
