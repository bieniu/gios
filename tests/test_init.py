"""Tests for gios package."""
import json

import aiohttp
import pytest
from aioresponses import aioresponses

from gios import ApiError, Gios, InvalidSensorsDataError, NoStationError

INVALID_STATION_ID = 0

VALID_STATION_ID = 552
VALID_STATION_NAME = "Test Name"
VALID_LATITUDE = 99.99
VALID_LONGITUDE = 88.88


@pytest.mark.asyncio
async def test_valid_data_first_value():
    """Test with valid data and valid first sensor's value."""
    with open("tests/fixtures/stations.json", encoding="utf-8") as file:
        stations = json.load(file)
    with open("tests/fixtures/station.json", encoding="utf-8") as file:
        station = json.load(file)
    with open("tests/fixtures/sensor_658.json", encoding="utf-8") as file:
        sensor_658 = json.load(file)
    with open("tests/fixtures/sensor_660.json", encoding="utf-8") as file:
        sensor_660 = json.load(file)
    with open("tests/fixtures/sensor_665.json", encoding="utf-8") as file:
        sensor_665 = json.load(file)
    with open("tests/fixtures/sensor_667.json", encoding="utf-8") as file:
        sensor_667 = json.load(file)
    with open("tests/fixtures/sensor_670.json", encoding="utf-8") as file:
        sensor_670 = json.load(file)
    with open("tests/fixtures/sensor_672.json", encoding="utf-8") as file:
        sensor_672 = json.load(file)
    with open("tests/fixtures/sensor_14395.json", encoding="utf-8") as file:
        sensor_14395 = json.load(file)
    with open("tests/fixtures/indexes.json", encoding="utf-8") as file:
        indexes = json.load(file)

    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/station/findAll",
            payload=stations,
        )
        session_mock.get(
            f"http://api.gios.gov.pl/pjp-api/rest/station/sensors/{VALID_STATION_ID}",
            payload=station,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/672",
            payload=sensor_672,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/658",
            payload=sensor_658,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/660",
            payload=sensor_660,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/665",
            payload=sensor_665,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/667",
            payload=sensor_667,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/670",
            payload=sensor_670,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/14395",
            payload=sensor_14395,
        )
        session_mock.get(
            f"http://api.gios.gov.pl/pjp-api/rest/aqindex/getIndex/{VALID_STATION_ID}",
            payload=indexes,
        )

        gios = Gios(VALID_STATION_ID, session)
        data = await gios.async_update()

    await session.close()

    assert gios.station_name == VALID_STATION_NAME
    assert gios.station_id == VALID_STATION_ID
    assert gios.latitude == VALID_LATITUDE
    assert gios.longitude == VALID_LONGITUDE
    assert data.so2.value == 11.6502
    assert data.so2.index == "very_good"
    assert data.c6h6.value == 2.57148
    assert data.c6h6.index == "very_good"
    assert data.co.value == 786.702
    assert data.co.index == "very_bad"
    assert data.no2.value == 59.9545
    assert data.no2.index == "very_good"
    assert data.o3.value == 8.63111
    assert data.o3.index == "bad"
    assert data.pm25.value == 59.9428
    assert data.pm25.index == "sufficient"
    assert data.pm10.value == 123.879
    assert data.pm10.index == "moderate"
    assert data.aqi.value == "good"


@pytest.mark.asyncio
async def test_api_error():
    """Test GIOS API error."""
    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/station/findAll",
            status=404,
        )
        gios = Gios(VALID_STATION_ID, session)

        with pytest.raises(ApiError) as excinfo:
            await gios.async_update()

        assert str(excinfo.value) == "404"

    await session.close()


@pytest.mark.asyncio
async def test_valid_data_second_value():
    """Test with valid data and valid second sensor's value."""
    with open("tests/fixtures/stations.json", encoding="utf-8") as file:
        stations = json.load(file)
    with open("tests/fixtures/station.json", encoding="utf-8") as file:
        station = json.load(file)
    with open("tests/fixtures/sensor_658.json", encoding="utf-8") as file:
        sensor_658 = json.load(file)
    with open("tests/fixtures/sensor_660.json", encoding="utf-8") as file:
        sensor_660 = json.load(file)
    with open("tests/fixtures/sensor_665.json", encoding="utf-8") as file:
        sensor_665 = json.load(file)
    with open("tests/fixtures/sensor_667.json", encoding="utf-8") as file:
        sensor_667 = json.load(file)
    with open("tests/fixtures/sensor_670.json", encoding="utf-8") as file:
        sensor_670 = json.load(file)
    with open("tests/fixtures/sensor_672.json", encoding="utf-8") as file:
        sensor_672 = json.load(file)
    with open("tests/fixtures/sensor_14395.json", encoding="utf-8") as file:
        sensor_14395 = json.load(file)
    with open("tests/fixtures/indexes.json", encoding="utf-8") as file:
        indexes = json.load(file)

    sensor_658["values"][0]["value"] = None
    sensor_660["values"][0]["value"] = None
    sensor_665["values"][0]["value"] = None
    sensor_667["values"][0]["value"] = None
    sensor_670["values"][0]["value"] = None
    sensor_672["values"][0]["value"] = None
    sensor_14395["values"][0]["value"] = None

    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/station/findAll",
            payload=stations,
        )
        session_mock.get(
            f"http://api.gios.gov.pl/pjp-api/rest/station/sensors/{VALID_STATION_ID}",
            payload=station,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/672",
            payload=sensor_672,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/658",
            payload=sensor_658,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/660",
            payload=sensor_660,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/665",
            payload=sensor_665,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/667",
            payload=sensor_667,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/670",
            payload=sensor_670,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/14395",
            payload=sensor_14395,
        )
        session_mock.get(
            f"http://api.gios.gov.pl/pjp-api/rest/aqindex/getIndex/{VALID_STATION_ID}",
            payload=indexes,
        )

        gios = Gios(VALID_STATION_ID, session)
        data = await gios.async_update()

    await session.close()

    assert gios.station_name == VALID_STATION_NAME
    assert gios.station_id == VALID_STATION_ID
    assert gios.latitude == VALID_LATITUDE
    assert gios.longitude == VALID_LONGITUDE
    assert data.so2.value == 11.501
    assert data.so2.index == "very_good"
    assert data.c6h6.value == 3.24432
    assert data.c6h6.index == "very_good"
    assert data.co.value == 1041.74
    assert data.co.index == "very_bad"
    assert data.no2.value == 52.6198
    assert data.no2.index == "very_good"
    assert data.o3.value == 4.93778
    assert data.o3.index == "bad"
    assert data.pm25.value == 72.0243
    assert data.pm25.index == "sufficient"
    assert data.pm10.value == 115.559
    assert data.pm10.index == "moderate"
    assert data.aqi.value == "good"


@pytest.mark.asyncio
async def test_no_indexes_data():
    """Test with valid data."""
    with open("tests/fixtures/stations.json", encoding="utf-8") as file:
        stations = json.load(file)
    with open("tests/fixtures/station.json", encoding="utf-8") as file:
        station = json.load(file)
    with open("tests/fixtures/sensor_658.json", encoding="utf-8") as file:
        sensor_658 = json.load(file)
    with open("tests/fixtures/sensor_660.json", encoding="utf-8") as file:
        sensor_660 = json.load(file)
    with open("tests/fixtures/sensor_665.json", encoding="utf-8") as file:
        sensor_665 = json.load(file)
    with open("tests/fixtures/sensor_667.json", encoding="utf-8") as file:
        sensor_667 = json.load(file)
    with open("tests/fixtures/sensor_670.json", encoding="utf-8") as file:
        sensor_670 = json.load(file)
    with open("tests/fixtures/sensor_672.json", encoding="utf-8") as file:
        sensor_672 = json.load(file)
    with open("tests/fixtures/sensor_14395.json", encoding="utf-8") as file:
        sensor_14395 = json.load(file)

    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/station/findAll",
            payload=stations,
        )
        session_mock.get(
            f"http://api.gios.gov.pl/pjp-api/rest/station/sensors/{VALID_STATION_ID}",
            payload=station,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/672",
            payload=sensor_672,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/658",
            payload=sensor_658,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/660",
            payload=sensor_660,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/665",
            payload=sensor_665,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/667",
            payload=sensor_667,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/670",
            payload=sensor_670,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/14395",
            payload=sensor_14395,
        )
        session_mock.get(
            f"http://api.gios.gov.pl/pjp-api/rest/aqindex/getIndex/{VALID_STATION_ID}",
            payload={},
        )

        gios = Gios(VALID_STATION_ID, session)
        data = await gios.async_update()

    await session.close()

    assert gios.station_name == VALID_STATION_NAME
    assert gios.station_id == VALID_STATION_ID
    assert gios.latitude == VALID_LATITUDE
    assert gios.longitude == VALID_LONGITUDE
    assert data.so2.value == 11.6502
    assert data.so2.index is None
    assert data.c6h6.value == 2.57148
    assert data.c6h6.index is None
    assert data.co.value == 786.702
    assert data.co.index is None
    assert data.no2.value == 59.9545
    assert data.no2.index is None
    assert data.o3.value == 8.63111
    assert data.o3.index is None
    assert data.pm25.value == 59.9428
    assert data.pm25.index is None
    assert data.pm10.value == 123.879
    assert data.pm10.index is None
    assert data.aqi is None


@pytest.mark.asyncio
async def test_no_sensor_data_1():
    """Test with no sensor data."""
    with open("tests/fixtures/stations.json", encoding="utf-8") as file:
        stations = json.load(file)
    with open("tests/fixtures/station.json", encoding="utf-8") as file:
        station = json.load(file)
    with open("tests/fixtures/sensor_658.json", encoding="utf-8") as file:
        sensor_658 = json.load(file)
    with open("tests/fixtures/sensor_660.json", encoding="utf-8") as file:
        sensor_660 = json.load(file)
    with open("tests/fixtures/sensor_665.json", encoding="utf-8") as file:
        sensor_665 = json.load(file)
    with open("tests/fixtures/sensor_667.json", encoding="utf-8") as file:
        sensor_667 = json.load(file)
    with open("tests/fixtures/sensor_670.json", encoding="utf-8") as file:
        sensor_670 = json.load(file)
    with open("tests/fixtures/sensor_672.json", encoding="utf-8") as file:
        sensor_672 = json.load(file)
    with open("tests/fixtures/sensor_14395.json", encoding="utf-8") as file:
        sensor_14395 = json.load(file)
    with open("tests/fixtures/indexes.json", encoding="utf-8") as file:
        indexes = json.load(file)

    sensor_658["values"][0]["value"] = None
    sensor_658["values"][1]["value"] = None
    sensor_660["values"][0]["value"] = None
    sensor_660["values"][1]["value"] = None
    sensor_665["values"][0]["value"] = None
    sensor_665["values"][1]["value"] = None
    sensor_667["values"][0]["value"] = None
    sensor_667["values"][1]["value"] = None
    sensor_670["values"][0]["value"] = None
    sensor_670["values"][1]["value"] = None
    sensor_672["values"][0]["value"] = None
    sensor_672["values"][1]["value"] = None
    sensor_14395["values"][0]["value"] = None
    sensor_14395["values"][1]["value"] = None

    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/station/findAll",
            payload=stations,
        )
        session_mock.get(
            f"http://api.gios.gov.pl/pjp-api/rest/station/sensors/{VALID_STATION_ID}",
            payload=station,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/672",
            payload=sensor_672,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/658",
            payload=sensor_658,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/660",
            payload=sensor_660,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/665",
            payload=sensor_665,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/667",
            payload=sensor_667,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/670",
            payload=sensor_670,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/14395",
            payload=sensor_14395,
        )
        session_mock.get(
            f"http://api.gios.gov.pl/pjp-api/rest/aqindex/getIndex/{VALID_STATION_ID}",
            payload=indexes,
        )
        gios = Gios(VALID_STATION_ID, session)

        with pytest.raises(InvalidSensorsDataError):
            await gios.async_update()

    await session.close()


@pytest.mark.asyncio
async def test_invalid_sensor_data_2():
    """Test with invalid sensor data."""
    with open("tests/fixtures/stations.json", encoding="utf-8") as file:
        stations = json.load(file)
    with open("tests/fixtures/station.json", encoding="utf-8") as file:
        station = json.load(file)

    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/station/findAll",
            payload=stations,
        )
        session_mock.get(
            f"http://api.gios.gov.pl/pjp-api/rest/station/sensors/{VALID_STATION_ID}",
            payload=station,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/672",
            payload=None,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/658",
            payload=None,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/660",
            payload=None,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/665",
            payload=None,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/667",
            payload=None,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/670",
            payload=None,
        )
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/data/getData/14395",
            payload=None,
        )
        gios = Gios(VALID_STATION_ID, session)

        with pytest.raises(InvalidSensorsDataError):
            await gios.async_update()

    await session.close()


@pytest.mark.asyncio
async def test_no_station_data():
    """Test with no station data."""
    with open("tests/fixtures/stations.json", encoding="utf-8") as file:
        stations = json.load(file)

    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/station/findAll",
            payload=stations,
        )
        session_mock.get(
            f"http://api.gios.gov.pl/pjp-api/rest/station/sensors/{VALID_STATION_ID}",
            payload={},
        )
        gios = Gios(VALID_STATION_ID, session)

        with pytest.raises(InvalidSensorsDataError):
            await gios.async_update()

    await session.close()


@pytest.mark.asyncio
async def test_no_stations_data():
    """Test with no stations data."""
    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/station/findAll",
            payload={},
        )
        gios = Gios(VALID_STATION_ID, session)

        with pytest.raises(ApiError):
            await gios.async_update()

    await session.close()


@pytest.mark.asyncio
async def test_invalid_station_id():
    """Test with invalid station_id."""
    with open("tests/fixtures/stations.json", encoding="utf-8") as file:
        stations = json.load(file)

    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://api.gios.gov.pl/pjp-api/rest/station/findAll",
            payload=stations,
        )
        gios = Gios(INVALID_STATION_ID, session)

        with pytest.raises(NoStationError) as excinfo:
            await gios.async_update()

        assert str(excinfo.value) == "0 is not a valid measuring station ID"

    await session.close()
