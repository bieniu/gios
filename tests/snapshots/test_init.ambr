# serializer version: 1
# name: test_no_indexes_data
  GiosSensors(aqi=None, c6h6=Sensor(name='benzene', id=658, index=None, value=2.57148), co=Sensor(name='carbon monoxide', id=660, index=None, value=786.702), no2=Sensor(name='nitrogen dioxide', id=665, index=None, value=59.9545), o3=Sensor(name='ozone', id=667, index=None, value=8.63111), pm10=Sensor(name='particulate matter 10', id=14395, index=None, value=123.879), pm25=Sensor(name='particulate matter 2.5', id=670, index=None, value=59.9428), so2=Sensor(name='sulfur dioxide', id=672, index=None, value=11.6502))
# ---
# name: test_valid_data_first_value
  GiosSensors(aqi=Sensor(name='AQI', id=None, index=None, value='good'), c6h6=Sensor(name='benzene', id=658, index='very_good', value=2.57148), co=Sensor(name='carbon monoxide', id=660, index='very_bad', value=786.702), no2=Sensor(name='nitrogen dioxide', id=665, index='very_good', value=59.9545), o3=Sensor(name='ozone', id=667, index='bad', value=8.63111), pm10=Sensor(name='particulate matter 10', id=14395, index='moderate', value=123.879), pm25=Sensor(name='particulate matter 2.5', id=670, index='sufficient', value=59.9428), so2=Sensor(name='sulfur dioxide', id=672, index='very_good', value=11.6502))
# ---
# name: test_valid_data_first_value.1
  GiosSensors(aqi=Sensor(name='AQI', id=None, index=None, value='good'), c6h6=Sensor(name='benzene', id=658, index='very_good', value=2.57148), co=Sensor(name='carbon monoxide', id=660, index='very_bad', value=786.702), no2=Sensor(name='nitrogen dioxide', id=665, index='very_good', value=59.9545), o3=Sensor(name='ozone', id=667, index='bad', value=8.63111), pm10=Sensor(name='particulate matter 10', id=14395, index='moderate', value=123.879), pm25=Sensor(name='particulate matter 2.5', id=670, index='sufficient', value=59.9428), so2=Sensor(name='sulfur dioxide', id=672, index='very_good', value=11.6502))
# ---
# name: test_valid_data_second_value
  GiosSensors(aqi=Sensor(name='AQI', id=None, index=None, value='good'), c6h6=Sensor(name='benzene', id=658, index='very_good', value=3.24432), co=Sensor(name='carbon monoxide', id=660, index='very_bad', value=1041.74), no2=Sensor(name='nitrogen dioxide', id=665, index='very_good', value=52.6198), o3=Sensor(name='ozone', id=667, index='bad', value=4.93778), pm10=Sensor(name='particulate matter 10', id=14395, index='moderate', value=115.559), pm25=Sensor(name='particulate matter 2.5', id=670, index='sufficient', value=72.0243), so2=Sensor(name='sulfur dioxide', id=672, index='very_good', value=11.501))
# ---
